### ISSUE TITLE
Specifying arguments on interfaces

### ISSUE THREAD

**Issue by samdenty** [MichalLytek/type-graphql#261]
The current `@Args()` (and Input) API only allows specifying arguments inside a resolver, which must be on an `ObjectType`.

Without going under the hood, it's impossible to specify arguments on an `InterfaceType`

Example (not possible - but valid GraphQL):

```ts
@InterfaceType()
class Interface {
  @Field()
  @Args(type => ArgsType)
  test: string
}
```

**Comment by 19majkel94** [comment]
Again, I've always treated GraphQL interface like a normal TS interface, that is pure abstract and doesn't have an implementation. So most people use it only as an constraint about a set of field rather than a signature of a method (field resolver). Looks like you are the first one after a year that need that feature 😉 

I agree that it's possible in GraphQL SDL. The ideal API would use abstract methods declarations:
```ts
@InterfaceType()
abstract class InterfaceWithArgs {
  @Field()
  abstract interfaceStringFieldWithArgs(
    @Arg("interfaceStringArg") interfaceStringArg: string,
  ): string;
}
```

But it gives me this error:
![image](https://user-images.githubusercontent.com/10618781/53288440-48532800-3788-11e9-89a5-07a3fc3ac686.png)

So it would require to create a normal method that just throw anything (null or `new AbstractMethod()`):

```ts
@InterfaceType()
abstract class InterfaceWithArgs {
  @Field()
  interfaceStringFieldWithArgs(
    @Arg("interfaceStringArg") interfaceStringArg: string,
  ): string {
    throw null;
  }
}
```

I think that it would be better to take an advantage of TS concepts like inheritance rather than trying to mimic the GraphQL type system and it's limitation.

Maybe you can show a real world example of using field args in interfaces?

**Comment by samdenty** [comment]
> I agree that it's possible in GraphQL SDL. The ideal API would use abstract methods declarations

> that is pure abstract and doesn't have an implementation

Hmm, yeah I think they should be pure, using just a `@Args()` decorator. The implementation should be left up to the `ObjectType`'s that implement it.

```ts
@InterfaceType()
abstract class Interface {
  @Field()
  // Any object that implements this class, must have the same arguments (or a more refined type)
  @Args(type => ArgsType, { nullable: true })
  test: string
}

@ObjectType({ implements: Interface })
class Object {
  @Field()
  // Putting the Args here would be redundant, as it is already inferred from the resolver
  // @Args(type => ArgsType)
  test: string
}

@Resolver(of => Object)
class ObjectResolver {
  @FieldResolver()
  test(@Args() args: ArgsType) {
    
  }
}
```

---

> Maybe you can show a real world example of using field args in interfaces?

Consider something like the following:


```graphql
interface Media {
  height: Int!
  width: Int!
  url(size: Int!): String!
}

type GIF implements Media {
  height: Int!
  width: Int!
  url(size: Int!, framerate: Int): String!
}

type Video implements Media {
  height: Int!
  width: Int!
  url(size: Int!, format: String): String!
}

# Imagine loads more Media implementations

type Query {
  media: Media!
}
```

Example queries:

```graphql
query {
  media {
    // Get a generic url
    url(size: 100)
  }
}

query {
  media {
    ... on GIF {
      url(size: 100, framerate: 60)
    }
  }
}
```

Without this feature, you'd have to do:

```graphql
query {
  media {
    ... on Video {
      url(size: 100)
    }
    ... on GIF {
      url(size: 100) // Need to repeat field arguments for every single Media implementation
    }
  }
}
```

**Comment by 19majkel94** [comment]
> using just a `@Args()` decorator

No, it's definitely against the TypeGraphQL design goals (1:1 type mapping between TS and GQL). It must have a method signature, not only a metadata decorator.

I have to think about the API for that, because the `throw null;` clausule would have to be repeated in `ObjectType` when you want to actually implement it if `FieldResolver`

> Need to repeat field arguments for every single Media implementation

You can use GraphQL union type to build a `Media` type and then you can place the field before the fragments.

**Comment by MichalLytek** [comment]
@samdenty 
Closed by #579, released in `v0.18.0-beta.14` 🎉 
Let me know how it works for you 😉
